### ISSUE TITLE
Enable For_Each Looping within Steampipe

### ISSUE THREAD

**Issue by Lachlan-White** [turbot/steampipe#3794]
**Is your feature request related to a problem? Please describe.**
In some instances, platforms have repetitious configurations required across multiple services. Such as with Azure we have Tagging, Diagnostic Settings and multiple more.

This is frustrating when trying to leverage HCL and the ease with which simple data modules can be iterated, which is common practice in other HCL-based tooling such as Terraform. Aligning to having similar capabilities would reduce the barrier to entry for users coming from other toolsets.

**Describe the solution you'd like**
A DRY-er way to do this would be to enable looping (for_each) through the resource types such as controls, queries etc. 

This would mean that rather than a control for every single Azure resource regarding mandatory tags, we can define a standard SQL Query, and then iterate through the control passing in parameters of resource name and table name.

This solution enables an easier way to define many controls and get value from the platform/tooling faster. As well as enabling a single point of update for the lifecycle of controls or tables. If the Plugins are to update table names a single place to update this in a data model is a much simpler change thane scowering hundreds of defined control blocks.

**Describe alternatives you've considered**
The only alternative is to be overly verbose and rewrite queries, where the difference between each is minute, leading to human error and updates to the lifecycle of queries/controls quite difficult. 

**Additional context**
A simple example of this in Terraform is linked:
https://rafaelmedeiros94.medium.com/terraform-creating-20-azure-virtual-machines-with-for-each-and-a-module-df01818f12bc

**Comment by e-gineer** [comment]
@Lachlan-White drafted the example below is Slack...

---------

Something like this, would mean you could define the resources, and tables in a single data structure but create multiple controls with the one config of a "control" object.

```
mod "mandatory_tags" {
  title = "Mandatory Tags"
}
variable "mandatory_tags" {
  type = list(string)
  description = "A list of the clients mandatory tags, in which to validate aginst their environment."
  default = ["Environment","Owner"]
}
locals {
  mandatory_sql_resource_group = replace(local.mandatory_sql, "__DIMENSIONS__", "resource_group, subscription_id")
}
locals {
  mandatory_sql = <<-EOQ
    with analysis as (
      select
        id,
        title,
        tags ?& $1 as has_mandatory_tags,
        to_jsonb($1) - array(select jsonb_object_keys(tags)) as missing_tags,
        __DIMENSIONS__
      from
        __TABLE_NAME__
    )
    select
      id as resource,
      case
        when has_mandatory_tags then 'ok'
        else 'alarm'
      end as status,
      case
        when has_mandatory_tags then title || ' has all mandatory tags.'
        else title || ' is missing tags: ' || array_to_string(array(select jsonb_array_elements_text(missing_tags)), ', ') || '.'
      end as reason,
      __DIMENSIONS__
    from
      analysis;
  EOQ
}
locals {
  azure_resource {
    azure_app_service_plan {
      resource = "Azure App Service"
      table_name = "azure_app_service_plan"
    },
    azure_batch_account {
      resource = "Azure Batch Account"
      table_name = "azure_batch_account"
    },
    container_registry {
      resource = "Azure Container Registry"
      table_name = "azure_container_registry"
    }
  }
}
control "mandatory_tags" {
  for_each = local.azure_resource[*]
  title = "${azure_resource.value.resource} should have mandatory tags"
  description = "Check if ${azure_resource.value.resource} have mandatory tags."
  sql = replace(local.mandatory_sql_resource_group, "__TABLE_NAME__", azure_resource.value.table_name)
  param "manadtory_tags" {
    defualt = var.manadtory_tags
  }
```

you could then also loop through the controls within the benchmark "children' rather than defining them all

**Comment by e-gineer** [comment]
Definitely a fan of for_each for certain scenarios in terraform and similar ... way better than the old count method :-)

In the case above though, I'm still not sure it's worth it? My thinking:

- The local.azure_resource block with the input data is still quite wordy and long. Minor improvement over repeating the control block, but not huge.
- We'd need to come up with a way to name the controls in this scenario ... since steampipe allows controls to be referenced from benchmarks etc that would require more complex use of splats, or referring to (dynamic) indexes, etc. Individual control definitions (while more verbose) are easier to reference by name - especially across mods.
- It's always a balance on capability vs readability. We feel the SQL command with the replace already makes these mods harder to work with, but overall is worth it because it's very repetitive and better for maintenance. Is the abstraction of the loop input data from the for_each worth it?

I do agree though ... if we were adding for_each in steampipe mods then it would definitely look like this and feel similar to terraform ... but overall I'm not sure it's worth it yet.

**Comment by e-gineer** [comment]
After some discussion internally we've decided to close this issue for now as we focus on other priorities. We'd definitely welcome more examples and use cases as they arise to consider.

TL;DR; "not now", not "no".

**Comment by Lachlan-White** [comment]
Hey @e-gineer thanks for the insights, 

1. The `local.azure_resource` block is wordy but centralised making the maintenance and changes easier and focused, thater than 1:many across many lines in potentially different files and missing them or making mistakes in the places for updates.

2. Terraform does this already, so maybe something you can grab from them. This is something used every day by millions of engineers so not something that compleicated, whilst some scenarios do get complicated with splats they are quite rare. 

3. I agree the SQL command with replace impacts readability and makes things more complicated, but the repetative nature is within a complicated form, rather than a simple data structure that users are more familiar with.

I think from an open-source point of view leaving it open for visibility to get input from the community would be better than closing it after internal discussions.
